File: JS/LinkedList/ReverseListII.js
// Definition for singly-linked list.
//			function Node(data){
//				this.data = data
//				this.next = null
//			}

module.exports = {
	//param A : head node of linked list
	//param B : integer
	//param C : integer
	//return the head node in the linked list
	reverseBetween : function(list, m, n){
        var prev = null, first = null, reverse = null, rslt = list;
        var i = 1;
        for(i = 1; i < m; i++){
            prev = list;
            list = list.next;
        }
        first = list;
        for(; i <= n; i++){
            var tmp = list.next;
            list.next = reverse;
            reverse = list;
            list = tmp;
        }
        if(prev != null)
            prev.next = reverse;
        first.next = list;
        if(prev == null)return reverse;
        return rslt;
	}
};

File: Java/LinkedList/ReverseListII.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     public int val;
 *     public ListNode next;
 *     ListNode(int x) { val = x; next = null; }
 * }
 */
public class Solution {
	public ListNode reverseBetween(ListNode a, int m, int n) {
	    ListNode reverse = null, prev = null, first = null, placeholder=a;
	    int i = 0;
	    for( i = 1; i < m;i++){
	        prev = a;
	        a = a.next;
	    }
	    first = a;
	    for(;i<=n;i++){
	        ListNode tmp = a.next;
	        a.next = reverse;
	        reverse = a;
	        a = tmp;
	    }
	    if(prev != null)
	    prev.next = reverse;
	    first.next = a;
	    if(m == 1)return reverse;
	    return placeholder;
	}
}
